File: litr-demo/src/main/java/com/linkedin/android/litr/demo/data/TransformationPresenter.java
/*
 * Copyright 2019 LinkedIn Corporation
 * All Rights Reserved.
 *
 * Licensed under the BSD 2-Clause License (the "License").  See License in the project root for
 * license information.
 */
package com.linkedin.android.litr.demo.data;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.PointF;
import android.media.MediaFormat;
import android.net.Uri;
import android.os.Build;
import android.text.TextUtils;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.linkedin.android.litr.MediaTransformer;
import com.linkedin.android.litr.MimeType;
import com.linkedin.android.litr.filter.GlFilter;
import com.linkedin.android.litr.utils.TransformationUtil;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class TransformationPresenter {

    private static final String TAG = TransformationPresenter.class.getSimpleName();

    protected static final String KEY_ROTATION = Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
            ? MediaFormat.KEY_ROTATION
            : "rotation-degrees";

    private final Context context;
    private final MediaTransformer mediaTransformer;

    public TransformationPresenter(@NonNull Context context,
                                   @NonNull MediaTransformer mediaTransformer) {
        this.context = context;
        this.mediaTransformer = mediaTransformer;
    }

    public void cancelTransformation(@NonNull String requestId) {
        mediaTransformer.cancel(requestId);
    }

    public void play(@Nullable Uri contentUri) {
        if (contentUri != null) {
            Intent playIntent = new Intent(Intent.ACTION_VIEW);
            playIntent.setDataAndType(contentUri, context.getContentResolver().getType(contentUri));
            playIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);

            context.startActivity(playIntent);
        }
    }

    @Nullable
    protected List<GlFilter> createGlFilters(@NonNull SourceMedia sourceMedia,
                                           @Nullable TargetVideoTrack targetTrack,
                                           float overlayWidth,
                                           @NonNull PointF position,
                                           float rotation) {
        List<GlFilter> glFilters = null;
        if (targetTrack != null && targetTrack.overlay != null) {
            try {
                Bitmap bitmap = BitmapFactory.decodeStream(context.getContentResolver().openInputStream(targetTrack.overlay));
                if (bitmap != null) {
                    float overlayHeight;
                    VideoTrackFormat sourceVideoTrackFormat = (VideoTrackFormat) sourceMedia.tracks.get(targetTrack.sourceTrackIndex);
                    if (sourceVideoTrackFormat.rotation == 90 || sourceVideoTrackFormat.rotation == 270) {
                        float overlayWidthPixels = overlayWidth * sourceVideoTrackFormat.height;
                        float overlayHeightPixels = overlayWidthPixels * bitmap.getHeight() / bitmap.getWidth();
                        overlayHeight = overlayHeightPixels / sourceVideoTrackFormat.width;
                    } else {
                        float overlayWidthPixels = overlayWidth * sourceVideoTrackFormat.width;
                        float overlayHeightPixels = overlayWidthPixels * bitmap.getHeight() / bitmap.getWidth();
                        overlayHeight = overlayHeightPixels / sourceVideoTrackFormat.height;
                    }

                    PointF size = new PointF(overlayWidth, overlayHeight);

                    GlFilter filter = TransformationUtil.createGlFilter(context,
                                                                        targetTrack.overlay,
                                                                        size,
                                                                        position,
                                                                        rotation);
                    if (filter != null) {
                        glFilters = new ArrayList<>();
                        glFilters.add(filter);
                    }
                }
            } catch (IOException ex) {
                Log.e(TAG, "Failed to extract audio track metadata: " + ex);
            }
        }

        return glFilters;
    }

    @Nullable
    protected MediaFormat createMediaFormat(@Nullable TargetTrack targetTrack) {
        MediaFormat mediaFormat = null;
        if (targetTrack != null && targetTrack.format != null) {
            if (targetTrack.format.mimeType.startsWith("video")) {
                mediaFormat = createVideoMediaFormat((VideoTrackFormat) targetTrack.format);
            } else if (targetTrack.format.mimeType.startsWith("audio")) {
                mediaFormat = createAudioMediaFormat((AudioTrackFormat) targetTrack.format);
            }
        }

        return mediaFormat;
    }

    @NonNull
    protected MediaFormat createVideoMediaFormat(@NonNull VideoTrackFormat trackFormat) {
        MediaFormat mediaFormat = new MediaFormat();
        mediaFormat.setString(MediaFormat.KEY_MIME, trackFormat.mimeType);
        mediaFormat.setInteger(MediaFormat.KEY_WIDTH, trackFormat.width);
        mediaFormat.setInteger(MediaFormat.KEY_HEIGHT, trackFormat.height);
        mediaFormat.setInteger(MediaFormat.KEY_BIT_RATE, trackFormat.bitrate);
        mediaFormat.setInteger(MediaFormat.KEY_I_FRAME_INTERVAL, trackFormat.keyFrameInterval);
        mediaFormat.setInteger(MediaFormat.KEY_FRAME_RATE, trackFormat.frameRate);
        mediaFormat.setLong(MediaFormat.KEY_DURATION, trackFormat.duration);
        mediaFormat.setInteger(KEY_ROTATION, trackFormat.rotation);

        return mediaFormat;
    }

    @NonNull
    protected MediaFormat createAudioMediaFormat(@NonNull AudioTrackFormat trackFormat) {
        MediaFormat mediaFormat = new MediaFormat();
        mediaFormat.setString(MediaFormat.KEY_MIME, trackFormat.mimeType);
        mediaFormat.setInteger(MediaFormat.KEY_CHANNEL_COUNT, trackFormat.channelCount);
        mediaFormat.setInteger(MediaFormat.KEY_SAMPLE_RATE, trackFormat.samplingRate);
        mediaFormat.setInteger(MediaFormat.KEY_BIT_RATE, trackFormat.bitrate);
        mediaFormat.setLong(MediaFormat.KEY_DURATION, trackFormat.duration);

        return mediaFormat;
    }

    protected boolean hasVp8OrVp9Track(@NonNull List<TargetTrack> targetTracks) {
        for (TargetTrack targetTrack : targetTracks) {
            if (!targetTrack.shouldInclude) {
                continue;
            }

            if (TextUtils.equals(targetTrack.format.mimeType, MimeType.VIDEO_VP8)
                    || TextUtils.equals(targetTrack.format.mimeType, MimeType.VIDEO_VP9)) {
                return true;
            }
        }
        return false;
    }

    protected boolean hasVp8OrVp9TrackFormat(@NonNull List<MediaTrackFormat> trackFormats) {
        for (MediaTrackFormat trackFormat : trackFormats) {
            if (TextUtils.equals(trackFormat.mimeType, MimeType.VIDEO_VP8)
                    || TextUtils.equals(trackFormat.mimeType, MimeType.VIDEO_VP9)) {
                return true;
            }
        }
        return false;
    }
}

File: litr-demo/src/main/java/com/linkedin/android/litr/demo/data/TransformationPresenter.kt
/*
 * Copyright 2022 LinkedIn Corporation
 * All Rights Reserved.
 *
 * Licensed under the BSD 2-Clause License (the "License").  See License in the project root for
 * license information.
 */
package com.linkedin.android.litr.demo.data

import android.content.Context
import com.linkedin.android.litr.MediaTransformer
import android.content.Intent
import android.graphics.PointF
import com.linkedin.android.litr.filter.GlFilter
import android.graphics.BitmapFactory
import com.linkedin.android.litr.utils.TransformationUtil
import android.media.MediaFormat
import android.net.Uri
import android.os.Build
import com.linkedin.android.litr.MimeType

val KEY_ROTATION =
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) MediaFormat.KEY_ROTATION else "rotation-degrees"

open class TransformationPresenter(
    private val context: Context,
    private val mediaTransformer: MediaTransformer
) {

    fun cancelTransformation(requestId: String) {
        mediaTransformer.cancel(requestId)
    }

    fun play(contentUri: Uri?) {
        if (contentUri != null) {
            val playIntent = Intent(Intent.ACTION_VIEW)
                .apply {
                    setDataAndType(contentUri, context.contentResolver.getType(contentUri))
                    addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                }
            context.startActivity(playIntent)
        }
    }

    protected fun createGlFilters(
        sourceMedia: SourceMedia,
        targetTrack: TargetVideoTrack?,
        overlayWidth: Float,
        position: PointF,
        rotation: Float
    ): List<GlFilter>? {
        return targetTrack?.overlay?.let { _ ->
            BitmapFactory.decodeStream(context.contentResolver.openInputStream(targetTrack.overlay))?.let { bitmap ->
                val sourceVideoTrackFormat = sourceMedia.tracks[targetTrack.sourceTrackIndex] as VideoTrackFormat
                val overlayHeight =
                    if (sourceVideoTrackFormat.rotation == 90 || sourceVideoTrackFormat.rotation == 270) {
                        val overlayWidthPixels = overlayWidth * sourceVideoTrackFormat.height
                        val overlayHeightPixels =
                            overlayWidthPixels * bitmap.height / bitmap.width
                        overlayHeightPixels / sourceVideoTrackFormat.width
                    } else {
                        val overlayWidthPixels = overlayWidth * sourceVideoTrackFormat.width
                        val overlayHeightPixels =
                            overlayWidthPixels * bitmap.height / bitmap.width
                        overlayHeightPixels / sourceVideoTrackFormat.height
                    }

                TransformationUtil.createGlFilter(
                    context,
                    targetTrack.overlay,
                    PointF(overlayWidth, overlayHeight),
                    position,
                    rotation
                )?.let {
                    listOf(it)
                }
            }
        }
    }

    protected fun createMediaFormat(targetTrack: TargetTrack?): MediaFormat? {
        return targetTrack?.format?.let { format ->
            if (format.mimeType.startsWith("video")) {
                createVideoMediaFormat(targetTrack.format as VideoTrackFormat)
            } else if (format.mimeType.startsWith("audio")) {
                createAudioMediaFormat(targetTrack.format as AudioTrackFormat)
            } else {
                null
            }
        }
    }

    protected fun createVideoMediaFormat(trackFormat: VideoTrackFormat): MediaFormat {
        return MediaFormat.createVideoFormat(trackFormat.mimeType, trackFormat.width, trackFormat.height)
            .apply {
                setInteger(MediaFormat.KEY_BIT_RATE, trackFormat.bitrate)
                setInteger(MediaFormat.KEY_I_FRAME_INTERVAL, trackFormat.keyFrameInterval)
                setInteger(MediaFormat.KEY_FRAME_RATE, trackFormat.frameRate)
                setLong(MediaFormat.KEY_DURATION, trackFormat.duration)
                setInteger(KEY_ROTATION, trackFormat.rotation)
            }
    }

    protected fun createAudioMediaFormat(trackFormat: AudioTrackFormat): MediaFormat {
        return MediaFormat.createAudioFormat(trackFormat.mimeType, trackFormat.samplingRate, trackFormat.channelCount)
            .apply {
                setInteger(MediaFormat.KEY_BIT_RATE, trackFormat.bitrate)
                setLong(MediaFormat.KEY_DURATION, trackFormat.duration)
            }
    }

    protected fun hasVp8OrVp9Track(targetTracks: List<TargetTrack>): Boolean {
        return targetTracks.any {
            it.shouldInclude &&
                (it.format.mimeType == MimeType.VIDEO_VP8 || it.format.mimeType == MimeType.VIDEO_VP9)
        }
    }

    protected fun hasVp8OrVp9TrackFormat(trackFormats: List<MediaTrackFormat>): Boolean {
        return trackFormats.any {
            it.mimeType == MimeType.VIDEO_VP8 || it.mimeType == MimeType.VIDEO_VP9
        }
    }
}
